src/tuntap/macosx/c_tuntap_macosx_obj.cpp

```diff
@@ -1,41 +1,148 @@
+    buf_ptr[3] = 0x1E;
+
+    //size_t write_bytes = m_stream_handle_ptr->write_some(boost::asio::buffer(buf, count), ec);
+    size_t write_bytes = write(*m_stream_handle_ptr.get(), boost::asio::buffer(buf, count), ec);
+    if (ec) throw std::runtime_error("write to TUN error: " + ec.message());
+    return write_bytes;
+}
+*/
+	return m_tun_stream.write_some(boost::asio::buffer(data, size));
 }
 
 size_t c_tuntap_macosx_obj::read_from_tun(unsigned char * const data, size_t size) {
-	_NOTREADY();
-	_UNUSED(data);
-	_UNUSED(size);
-	return 0;
+/* OLD STYLE todo - waiting for memory model
+size_t c_tun_device_apple::read_from_tun(void *buf, size_t count) {
+    assert(m_readed_bytes > 0);
+    if(m_readed_bytes > count) throw std::runtime_error("undersized buffer");
+    // TUN header
+    m_buffer[0] = 0x00;
+    m_buffer[1] = 0x00;
+    m_buffer[2] = 0x86;
+    m_buffer[3] = 0xDD;
+    std::copy_n(&m_buffer[0], m_readed_bytes, static_cast<uint8_t *>(buf));
+    size_t ret = m_readed_bytes;
+    m_readed_bytes = 0;
+    return ret;
+}
+*/
+	return m_tun_stream.read_some(boost::asio::buffer(data, size));
 }
 
 void c_tuntap_macosx_obj::async_receive_from_tun(unsigned char * const data,
                                                  size_t size,
                                                  const c_tuntap_base_obj::read_handler &handler) {
-	_NOTREADY();
-	_UNUSED(data);
-	_UNUSED(size);
-	_UNUSED(handler);
+
+	auto asio_handler = [data, handler](const boost::system::error_code& error, std::size_t bytes_transferred) {
+		handler(data, bytes_transferred, error);
+	};
+	return m_tun_stream.async_read_some(boost::asio::buffer(data, size), asio_handler);
 }
 
-void c_tuntap_macosx_obj::set_tun_parameters(const std::array<unsigned char, 16> &binary_address,
+void c_tuntap_macosx_obj::set_tun_parameters(const std::array<unsigned char, IPV6_LEN> &binary_address,
                                              int prefix_len,
                                              uint32_t mtu) {
-	_NOTREADY();
-	_UNUSED(binary_address);
-	_UNUSED(prefix_len);
-	_UNUSED(mtu);
+
+	c_haship_addr address(c_haship_addr::tag_constr_by_array_uchar(), binary_address);
+	_goal("Configuring tuntap options: IP address: " << address << "/" << prefix_len << " MTU=" << mtu);
+
+	set_ipv6_address(binary_address, prefix_len);
+	set_mtu(mtu);
+
+	m_tun_stream.release();
+	m_tun_stream.assign(m_tun_fd);
+	_goal("Configuring tuntap options - done");
 }
 
-#endif // ANTINET_macosx
+int c_tuntap_macosx_obj::create_tun_fd() {
+	int tun_fd = socket(PF_SYSTEM, SOCK_DGRAM, SYSPROTO_CONTROL);
+	int err=errno;
+	if (tun_fd < 0)
+		_throw_error_sub(tuntap_error_devtun,
+		                 NetPlatform_syserr_to_string({e_netplatform_err_open_fd, err}) );
 
+	// get ctl_id
+	ctl_info info;
+	std::memset(&info, 0, sizeof(info));
+	const std::string apple_utun_control = "com.apple.net.utun_control";
+	apple_utun_control.copy(info.ctl_name, apple_utun_control.size());
```

Comment:
to assert against too long strings,

.copy(info.ctl_name,  sizeof( info.ctl_name )    ));
also static_assert that info.ctl_name  is char / is sizeof==1